samples/tcp/gopher.js

```diff
@@ -32,19 +29,11 @@ export default {
 
       // The Gopher server should now return a response to us and close the connection.
       // So start reading from the socket.
-      const reader = socket.readable.getReader();
       const decoder = new TextDecoder();
 
       let gopherResponse = "";
-      while (true) {
-        const res = await reader.read();
-        if (res.done) {
-          console.log("Stream done, socket connection has been closed.");
-          // TODO(soon): Half-open connections:
-          // see https://github.com/cloudflare/workerd/pull/162#discussion_r1020483277.
-          break;
-        }
-        gopherResponse += decoder.decode(res.value);
+      for await (const chunk of socket.readable) {
+        gopherResponse += decoder.decode(chunk);
       }
```

Comment:
For decoding, you'll want...

```suggestion
      for await (const chunk of socket.readable) {
        gopherResponse += decoder.decode(chunk, { stream: true });
      }
      gopherResponse += decoder.decode(chunk);
```

This ensures that if the chunk cuts off any multi-byte characters they'll be properly handled.

Reply:
I assume that last `gopherResponse += decoder.decode(chunk);` line is a mistake

Reply:
Doh, yes, it should be `gopherResponse += decoder.decode()` with no chunk. It's necessary to flush any remaining items from the decoders buffer.

Reply:
```suggestion
      for await (const chunk of socket.readable) {
        gopherResponse += decoder.decode(chunk, { stream: true });
      }
      gopherResponse += decoder.decode();
```

Reply:
or is it? but then how does this work? I would expect the `chunk` to be only accessible inside the for loop.

Reply:
Surprisingly it worked. Does the `chunk` exist outside the `for` loop scope?